rust/diesel/src/order.rs

```diff
@@ -0,0 +1,163 @@
+use bigdecimal::BigDecimal;
+use diesel;
+use diesel::pg::PgConnection;
+use diesel::prelude::*;
+use num_traits::identities::Zero;
+use serde::{Deserialize, Serialize};
+use std::ops::{Add, Mul};
+
+use crate::product::Product;
+use crate::schema::order_lines;
+use crate::schema::orders;
+
+#[derive(AsChangeset, Serialize, Deserialize, Queryable, Insertable)]
+#[serde(rename_all = "camelCase")]
+pub struct Order {
+    pub order_id: i32,
+    pub order_total: BigDecimal,
+    pub user_id: i32,
+}
+
+#[table_name = "orders"]
+#[derive(AsChangeset, Serialize, Deserialize, Queryable, Insertable)]
+#[serde(rename_all = "camelCase")]
+struct NewOrder {
+    pub order_total: BigDecimal,
+    pub user_id: i32,
+}
+
+#[table_name = "order_lines"]
+#[derive(AsChangeset, Serialize, Deserialize, Queryable, Insertable)]
+#[serde(rename_all = "camelCase")]
+pub struct NewOrderLine {
+    pub order_id: i32,
+    pub product_id: i32,
+    pub units: i16,
+}
+
+#[derive(Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct NewUserOrder {
+    pub user_id: i32,
+    pub products: Vec<ProductOrder>,
+}
+
+#[derive(Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct ProductOrder {
+    pub product_id: i32,
+    pub units: i16,
+}
+
+impl Order {
+    pub fn create(order: NewUserOrder, connection: &PgConnection) -> Option<Order> {
+        let (row, product_units) = Self::create_new_order(&order, connection)?;
+
+        let insert_result = connection
+            .build_transaction()
+            .serializable()
+            .read_write()
+            .run(|| {
+                let inserted_order: Order = diesel::insert_into(orders::table)
+                    .values(&row)
+                    .get_result(connection)?;
+
+                let order_lines: Vec<NewOrderLine> = product_units
```

Comment:
I think you should also be able to extract this and the following calls into their own method, since this is _almost_ exactly duplicated between the create and update paths. Perhaps something like:
```
fn createOrderLines(product_units, order, connection) -> Result<...> {...}
```